Feature/adl 36

S3784866 opened the pull request:
Footer bar
- Displays CarID and UTC Time from AV Logger.
- Created basic Heartbeat message relay and displays alongside connection.
- Raw AV Data logger messages displayed in Console box with auto scroll function.

S3784866 committed:
Added base footer UI design and CarID and UTC gathering from AV Logger sim

S3784866 committed:
Got the Time and Raw data message updates to work, created RawData to hold the raw message, CarID and UTC time.

S3784866 committed:
Got the 'Heartbeat' and 'Connection' indicators somewhat working

S3784866 committed:
Merge branch 'main' of https://github.com/nickowen44/AV-Datalogger into feature/ADL-36 in prep for PR.

S3784866 committed:
Got Heartbeat and Connection actually working

S3784866 committed:
Actually get Heartbeat/Connection working 2.0

S3784866 committed:
Actually get Heartbeat/Connection working 3.0

S3784866 committed:
Cleaning up

S3784866 committed:
Final clean before PR

S3784866 committed:
Removed grid lines

S3784866 committed:
Trying to fix test build issues

S3784866 committed:
Removed time conversion

nrdag commented on Dashboard/Utils/DependencyInjection.cs:
```diff
@@ -19,11 +19,13 @@ public static IServiceProvider ConfigureServices()
 
         services.AddTransient<DataViewModel>();
         services.AddTransient<MainWindowViewModel>();
         services.AddTransient<ISerialPort, SerialPortWrapper>();
```
These are doubled up now, i think i put this as transient so we can potentially recreate this at runtime via the connection page, but not quite sure how that would work actually

S3784866 replied:
I didn't even realized I had done this, I'll remove it.

nrdag commented on Dashboard/ViewModels/FooterViewModel.cs:
```diff
@@ -0,0 +1,98 @@
+    public bool ConnectionStatus => _dataStore.RawData?.ConnectionStatus ?? false;
+    public string CarId => _dataStore.RawData?.CarId ?? "0";
+    public string RawMessage => _dataStore.RawData?.RawMessage ?? "";
+    public string UTcTime
```
Change var name to UTCTime plz

S3784866 replied:
Will do.
https://github.com/nickowen44/AV-Datalogger/pull/10/commits/dfa7244d705bf6043b7764f4c730fbe36a0eb0a2

nrdag commented on Dashboard/ViewModels/FooterViewModel.cs:
```diff
@@ -0,0 +1,98 @@
+    {
+        _dataStore.RawDataUpdated -= OnRawDataChanged;
+        _dataStore.HeartBeatUpdated -= OnHeartbeatStatusChanged;
+        // _dataStore.Dispose();
```
plz remove commented code

S3784866 replied:
Will do.
https://github.com/nickowen44/AV-Datalogger/pull/10/commits/dfa7244d705bf6043b7764f4c730fbe36a0eb0a2

nrdag commented on Dashboard/Models/DataStore.cs:
```diff
@@ -8,10 +8,14 @@ public class DataStore : IDataStore, IDisposable
     public event EventHandler? GpsDataUpdated;
     public event EventHandler? AvDataUpdated;
     public event EventHandler? ResDataUpdated;
+    public event EventHandler? RawDataUpdated;
```
I am cautious about adding this, I am unsure if we want to just straight output everything via events, it may be nicer if we do this via a log file, and either have some log event that gets consumed by our view model or we tail a log file for the console?

S3784866 replied:
I'm also not too sure about this, I think it should be fine for right now and maybe we can loop back when we get the logging done?

nrdag replied:
That works for me, can you add to the AC for the logging ticket to do this? Maybe might need to re-estimate the ticket?

nrdag commented on Dashboard/ViewModels/FooterViewModel.cs:
```diff
@@ -0,0 +1,98 @@
+            {
+                return "Invalid time";
+            }
+            try
```
Could we do this conversion at the serial layer? So everything is consuming proper formatted times? (maybe convert to a datetime object and pass that around)

nrdag replied:
additional q: do we even need to pull this from the av logger? can we just do this ourselves and get the current time in C#?

S3784866 replied:
We could do this at the serial layer so other pages could use the times if needed.

As for not pulling it from the logger, I think it would depend on if we're making this in prep for replaying logs and if Nick O wants this UTC time to be the current or time of the original log.

nrdag replied:
Discussed in the meeting with nick yesterday, will use the time from the logger as it will allow more flexibility when we add replay from file

nrdag commented on Dashboard/Connectors/Serial/SerialConnector.cs:
```diff
@@ -10,6 +11,10 @@ public class SerialConnector(ISerialPort comPort) : IConnector
+    public event EventHandler<RawData>? RawDataUpdated;
+    public event EventHandler<bool>? HeartBeatUpdated;
+    private Timer? _heartbeatTimer;
+    private const string HeartBeatMessage = "HeartBeat";
```
According to the spec, this should just be `CON?`

nrdag replied:
And the logger should return `OK` to signal that a device is connected to the logger

nrdag replied:
are we meant to be heart beat checking PC -> AV logger, or AV logger -> car?

nrdag replied:
> are we meant to be heart beat checking PC -> AV logger, or AV logger -> car?

we should be checking pc -> logger (serial connection)

nrdag commented on Dashboard/Connectors/Serial/SerialConnector.cs:
```diff
@@ -19,6 +24,12 @@ public void Start()
         // Set up the event handler for when data is received
         comPort.DataReceived += OnDataReceived;
 
+        // Setting up timer for heatbeat messages
```
Do we need to be sending a heart beat every second, can we just check that the last received message was within some timeframe, and if not we send it?

nrdag replied:
According to Nick O, we should consider sending the heartbeat only when we haven't received a message in some predetermined amount of time

nrdag commented on Dashboard/Connectors/Serial/SerialConnector.cs:
```diff
@@ -38,18 +49,32 @@ public void Stop()
         // First we validate that we have the correct message format
         if (!message.StartsWith("#ID=") || !message.Contains("|UTC="))
             throw new InvalidOperationException("Invalid message format received", new Exception(message));
+        // Very dirty method right now, Need to ask Nick how to better handle the first message being out of format
```
I fixed this stuff in ADL-45, this is a reminder for me to ensure this is fixed post merge (whichever ticket goes first)

nrdag commented on Dashboard/Connectors/Serial/SerialConnector.cs:
```diff
@@ -161,4 +198,19 @@ private bool ParseBool(string value)
+        {
+
+            Console.WriteLine($"Sending heartbeat: {HeartBeatMessage}");
+            comPort.Write(HeartBeatMessage);
```
For now we could have some fake implementation that returns the `OK` message when we do a connection test?

nrdag replied:
so we can implement the handling of that now rather than later

nrdag commented on Dashboard/Connectors/Serial/SerialPortWrapper.cs:
```diff
@@ -65,4 +65,21 @@ public bool IsConnected
+            _serialPort.WriteTimeout = 50;
+            try
+            {
+                _serialPort.Write(data);
```
q: does this block the main thread? do we need to consider opening a thread that writes to the port also?

S3784866 replied:
Now that you've brought it up, I think it can block the main thread, I'll open a thread to handle this.

nrdag commented on Dashboard/ViewModels/MainWindowViewModel.cs:
```diff
@@ -39,10 +42,15 @@ public partial class MainWindowViewModel : ViewModelBase
             _serviceProvider = DependencyInjection.ConfigureServices();
             Items = new ObservableCollection<ListItemTemplate>(_templates);
             SelectedListItem = Items.First();
+            _footer = new FooterView
```
Does the footer change at runtime? Can we not just put the usercontrol directly in the MainWindowView.axaml XML?

S3784866 replied:
The Footer could be directly a part of MainWindowView but it would come down to if we want to keep the Navigation system and Footer logic separate for maintaining. I opt'ed to keep them separate just to air on the side of being safe on that regard.

nrdag replied:
I am thinking the view (or control) can remain separate, thats fine. my question was just can we do this statically via the XAML? so no need to dynamically create it in this file

S3784866 replied:
I don't think so since the footer needs its own DataContext if we keep them separate from my understanding.

nrdag replied:
Yep, I just tried it myself, I see what you mean. All good :)

nrdag commented on Dashboard/Views/MainWindowView.axaml:
```diff
@@ -51,5 +51,8 @@
         <ContentControl Name="MainContent" Content="{Binding CurrentPage}"/> 
     </Grid>
+        <Grid Grid.Row="2">
+            <ContentControl Name="FooterContent" Content="{Binding Footer}"/> 
```
can just be the footer directly here? <Footer>

S3784866 committed:
Addressed the following issues
- Remove commented code.
- Rename variables.
- Removed double up in Dependacy Injection.
- Moved Time convertion into serial port (ParseTime) in DateTime format.
- Moved Heartbeat into its own thread to not block the main thread and changed message into CON? with basic handling of AV Logger response.

S3784866 committed:
Fixed merge conflicts

S3784866 committed:
Fixing conflicts again

S3784866 committed:
Fixing FooterTest with local time convertion

S3784866 committed:
Fixed scaling of footer

nrdag commented on Dashboard/Connectors/Serial/SerialPortWrapper.cs:
```diff
@@ -76,4 +76,30 @@ public bool IsConnected
+                _serialPort.Write(data);
+
+                // If the AV Logger doesn't respond OK then connection is dead.
+                if (_serialPort.ReadLine() != "OK")
```
I don't think there is a guarantee that the next message read will be "OK", we should maybe add this as a message type to parse.

nrdag replied:
my meaning behind ^ is there is a potential case where there is another message currently in the buffer, given we run this multi threaded we can't guarantee the order of things

S3784866 replied:
Ya, I wasn't too sure how we were going to handle this. I don't think we're able to test it since the sim doesn't respond to commands

nrdag commented on Dashboard/Connectors/Serial/SerialConnector.cs:
```diff
@@ -23,6 +33,23 @@ public void Start()
+        {
+            while (_heartBeatShouldRun)
+            {
+                var timeSinceLastMessage = DateTime.Now - _lastMessageReceived;
```
There is tracking done within SerialPortWrapper -> IsConnected for this type of logic already, you could poll that instead. It currently checks if it has received any message within X seconds and if the port is open

S3784866 replied:
I can swap the check to use IsConnected, I don't know why I didn't use that in the first place considering I was using IsConnected before the threading.

S3784866 replied:
https://github.com/nickowen44/AV-Datalogger/pull/10/commits/af9d6f189027cdae3bff3d19679e23588c2190d4

nrdag commented on Dashboard/Connectors/Serial/SerialConnector.cs:
```diff
@@ -48,8 +80,7 @@ private void ParseMessage(string message)
 
-        // All messages start with the format "#ID=<ID>|UTC=<TIME>|<MSG>", so we split the message by the '|', and remove the first 2 elements
-        var split = message.Split('|')[2..];
+        var split = message.Substring(1).Split('|');
```
What does this change mean? Is it for parsing a potential "OK" message? Could consider adding a strict eq check before the regex match on L80

nrdag replied:
(same as change on the L94 and L98)

S3784866 replied:
1. It's to grab the ID and UTC time while trimming the first # off the ID.
2. I believe that was something I put in place when I brought up the buffer issue, you had asked to change the check to check for < 2 rather than != 2 as a stopgap. I can change it back to != 2.
3.  I changed it from using the Add function so it handles double up key-value pairs since we weren't checking for double ups. Currently I have it using the latest value of the key-value pair.

nrdag commented on Dashboard/Connectors/Serial/SerialConnector.cs:
```diff
@@ -143,6 +175,42 @@ private void ParseResMessage(Dictionary<string, string> values)
+    private DateTime ParseUTCTime(string timeString)
+    {
+
+        string datePart = "";
```
Good job on this, I don't like that it is required tbh. I think it would be best if we could push back and ask for an ISO 8601 str, but i think too much effort

S3784866 replied:
It would be nice to have it already be in a standard format. We can bring it up to Nick O next meeting if we have time.

S3784866 replied:
It would be good to have it already in a standard format, We could bring it up to Nick O but given the timeline we currently have, its not a big issue.

nrdag commented on Dashboard/Connectors/Serial/SerialConnector.cs:
```diff
@@ -143,6 +175,42 @@ private void ParseResMessage(Dictionary<string, string> values)
+
+        string timePart = timeString.Substring(9, 8);
+
+        DateTime parsedDate = DateTime.ParseExact(datePart, "yyyyMMdd", CultureInfo.InvariantCulture);
```
is there reasoning behind these being separate, im guessing due to formatting issues

S3784866 replied:
Ya, I was having format issues, although I think I know how to shorten this to just use one variable.

S3784866 replied:
https://github.com/nickowen44/AV-Datalogger/pull/10/commits/af9d6f189027cdae3bff3d19679e23588c2190d4

nrdag commented on Dashboard/Models/RawData.cs:
```diff
@@ -0,0 +1,11 @@
+
+public record RawData
+{
+    public string CarId { get; init; }
```
U can fix this warning by adjusting all (or just hte ones that warn) to look like
```suggestion
    public required string CarId { get; init; }
```

nrdag replied:
just means when the constructor or init statement is exited, it has to have a value

S3784866 replied:
Can do.

S3784866 replied:
https://github.com/nickowen44/AV-Datalogger/pull/10/commits/af9d6f189027cdae3bff3d19679e23588c2190d4

nrdag commented on Dashboard/Views/FooterView.axaml:
```diff
@@ -0,0 +1,69 @@
+    </Design.DataContext>
+    <Grid ColumnDefinitions="0.2*,0.2*, 0.6*" >
+        <Grid Grid.Column="0">
+            <StackPanel VerticalAlignment="Center">
```
Can you play around with the padding / margins plz:
- On the left of the times can you match the same margin that is to the right of hte console box
- Between the radios can you add a bit of padding

S3784866 replied:
Can do.

S3784866 replied:
https://github.com/nickowen44/AV-Datalogger/pull/10/commits/af9d6f189027cdae3bff3d19679e23588c2190d4

S3784866 committed:
Addressed PR comments
- Padding/Margins on content.
- Added required keyword to RawData record.
- Refactored ParseTime function to be shorter.
- Changed heartBeatThread to use IsConnected instead of manually checking timeout status

S3784866 committed:
Fixed merge conflicts

S3784866 committed:
Linting

S3784866 committed:
Added slight margin to Connection indicator

nrdag commented:
LGTM
